File: backend/app/utils/safe_url.py
import ipaddress
import socket
from urllib.parse import urlparse

from app.errors import AppError

BLOCKED_HOSTNAMES = {"localhost", "metadata.google.internal"}


def _is_private_ip(ip: str) -> bool:
    try:
        addr = ipaddress.ip_address(ip)
        return addr.is_private or addr.is_loopback or addr.is_link_local
    except ValueError:
        return False


async def assert_safe_public_url(raw_url: str):
    try:
        parsed = urlparse(raw_url)
        if not parsed.scheme or not parsed.netloc:
            raise ValueError("invalid")
    except ValueError as exc:
        raise AppError("Invalid URL format", 400) from exc

    if parsed.scheme not in ("http", "https"):
        raise AppError("Only HTTP and HTTPS URLs are allowed", 400)

    hostname = parsed.hostname
    if not hostname:
        raise AppError("Invalid URL format", 400)

    hostname_lower = hostname.lower()
    if hostname_lower in BLOCKED_HOSTNAMES:
        raise AppError("URL host is not allowed", 400)

    try:
        ipaddress.ip_address(hostname)
        if _is_private_ip(hostname):
            raise AppError("URL host is not allowed", 400)
        return parsed
    except ValueError:
        pass

    loop = socket.getaddrinfo(hostname, None)
    for entry in loop:
        addr = entry[4][0]
        if _is_private_ip(addr):
            raise AppError("URL host is not allowed", 400)

    return parsed

File: backend/src/utils/security/safeUrl.ts
import dns from "dns/promises";
import net from "net";
import { AppError } from "../../lib/errors/AppError";

const BLOCKED_HOSTNAMES = new Set([
  "localhost",
  "metadata.google.internal",
]);

function isPrivateIp(ip: string): boolean {
  if (net.isIPv4(ip)) {
    const parts = ip.split(".").map(Number);
    if (parts[0] === 10) return true;
    if (parts[0] === 127) return true;
    if (parts[0] === 169 && parts[1] === 254) return true;
    if (parts[0] === 172 && parts[1] >= 16 && parts[1] <= 31) return true;
    if (parts[0] === 192 && parts[1] === 168) return true;
    if (parts[0] === 0) return true;
  }
  if (net.isIPv6(ip)) {
    const normalized = ip.toLowerCase();
    if (normalized === "::1" || normalized.startsWith("fe80:") || normalized.startsWith("fc") || normalized.startsWith("fd")) {
      return true;
    }
  }
  return false;
}

/**
 * Validates URL for server-side fetch (SSRF mitigation).
 * Allows only http/https to public hosts; resolves DNS and rejects private IPs.
 */
export async function assertSafePublicUrl(rawUrl: string): Promise<URL> {
  let parsed: URL;
  try {
    parsed = new URL(rawUrl);
  } catch {
    throw new AppError("Invalid URL format", 400);
  }

  if (!["http:", "https:"].includes(parsed.protocol)) {
    throw new AppError("Only HTTP and HTTPS URLs are allowed", 400);
  }

  const hostname = parsed.hostname.toLowerCase();
  if (BLOCKED_HOSTNAMES.has(hostname)) {
    throw new AppError("URL host is not allowed", 400);
  }

  if (net.isIP(hostname)) {
    if (isPrivateIp(hostname)) {
      throw new AppError("URL host is not allowed", 400);
    }
    return parsed;
  }

  const addresses = await dns.lookup(hostname, { all: true });
  for (const entry of addresses) {
    if (isPrivateIp(entry.address)) {
      throw new AppError("URL host is not allowed", 400);
    }
  }

  return parsed;
}

export default assertSafePublicUrl;
